File: android/overlay/common/src/commonMain/kotlin/chat/hearth/HearthRelayPort.kt
package chat.hearth

/**
 * Порт, на котором телефоны ходят к своему релею.
 *
 * # Почему 443, а не 5223
 *
 * 5223 — родной порт SMP, но нестандартный, и его режут офисный и публичный Wi-Fi и
 * часть провайдеров: соединение пропускают, а первое сообщение шифрования глушат.
 * Именно так телефон «не мог создать ссылку» — на узле было видно, что соединение
 * открывается, а до сервера не доходит ни байта, и через 60 секунд сервер его
 * закрывает. 443 — порт обычных сайтов, его не режет почти никто.
 *
 * Серверы SimpleX клиент и сам водит через 443 (`smpWebPortServers = preset` в ядре);
 * поэтому ссылки на их серверах из той же сети создавались, а на наших — нет.
 *
 * Релей слушает оба порта, так что переписывание ничего не ломает: уже созданные
 * очереди продолжают жить на 5223, новые идут через 443.
 */
const val HEARTH_RELAY_WEB_PORT = 443
private const val SMP_DEFAULT_PORT = 5223

// smp://<отпечаток>:<пароль>@<хост[,хост…]>[:порт]
private val SMP_ADDRESS = Regex("^(smp://[^@/?# ]+@)([A-Za-z0-9.,-]+)(?::([0-9]{1,5}))?$")

/**
 * Адрес своего релея, переведённый на 443, или `null`, если трогать нечего.
 *
 * Переписываем только SMP, только свой хост и только порт 5223 — явный или
 * подразумеваемый. Чужие серверы, XFTP (у него свой порт) и нестандартные порты,
 * которые кто-то задал сознательно, не трогаем.
 */
fun hearthRelayAddressOnWebPort(address: String, relayHost: String): String? {
  val m = SMP_ADDRESS.matchEntire(address.trim()) ?: return null
  val (prefix, hosts, port) = m.destructured
  if (hosts.split(',').none { it.equals(relayHost, ignoreCase = true) }) return null
  val current = port.toIntOrNull() ?: SMP_DEFAULT_PORT
  if (current != SMP_DEFAULT_PORT) return null
  return "$prefix$hosts:$HEARTH_RELAY_WEB_PORT"
}

File: ios/overlay/SimpleXChat/Hearth/HearthRelayPort.swift
//
//  HearthRelayPort.swift
//  Hearth
//
//  Порт, на котором телефоны ходят к своему релею.
//

import Foundation

/// # Почему 8443, а не 443 и не 5223
///
/// Домашний провайдер узла на входе досматривает порты 443 и 5223 и выбрасывает именно
/// TLS-приветствие клиента SimpleX. Тот же пакет на любой другой порт доходит; проверено
/// на двадцати портах (`docs/deploy-fels-2026-09-09.md`). 8443 — обычный «альтернативный
/// HTTPS», его пропускают и офисные сети. Релей слушает и старые порты, так что
/// переписывание ничего не ломает.
public enum HearthRelayPort {
    public static let webPort = 8443
    static let smpDefaultPort = 5223
    /// Порты, с которых переводим: родной SMP и тот, что выдавался до обнаружения DPI.
    static let moveFrom: Set<Int> = [smpDefaultPort, 443]

    // smp://<отпечаток>:<пароль>@<хост[,хост…]>[:порт]
    private static let pattern = try? NSRegularExpression(pattern: "^(smp://[^@/?# ]+@)([A-Za-z0-9.,-]+)(?::([0-9]{1,5}))?$")

    /// Адрес своего релея, переведённый на 8443, или `nil`, если трогать нечего.
    ///
    /// Переписываем только SMP, только свой хост и только порты 5223 (явный или
    /// подразумеваемый) и 443. Чужие серверы, XFTP (у него свой порт) и нестандартные
    /// порты, которые кто-то задал сознательно, не трогаем.
    public static func migrate(_ address: String, ownHost: String) -> String? {
        let trimmed = address.trimmingCharacters(in: .whitespacesAndNewlines)
        let ns = trimmed as NSString
        let full = NSRange(location: 0, length: ns.length)
        guard let regex = pattern,
              let m = regex.firstMatch(in: trimmed, options: [], range: full),
              m.range == full
        else { return nil }
        let prefix = ns.substring(with: m.range(at: 1))
        let hosts = ns.substring(with: m.range(at: 2))
        guard hosts.split(separator: ",", omittingEmptySubsequences: false).contains(where: { HearthText.sameHost($0, ownHost) }) else {
            return nil
        }
        let portRange = m.range(at: 3)
        let current = portRange.location == NSNotFound ? smpDefaultPort : (Int(ns.substring(with: portRange)) ?? -1)
        guard moveFrom.contains(current) else { return nil }
        return "\(prefix)\(hosts):\(webPort)"
    }
}
